File: src/abstractClient.ts
import { io } from "socket.io-client";
import { ConsoleReader } from "./consoleReader";
import { IdleTimer } from "./idleTimer";

const SOCKET_ADDRESS = "http://localhost:3001";
const TIMEOUT = 10 * 1000 * 60;

export abstract class AbstractClient {
  protected socket = io(SOCKET_ADDRESS);
  protected consoleReader = new ConsoleReader();
  protected idleTimer = new IdleTimer(() => process.exit(), TIMEOUT);

  abstract start: () => Promise<void>;
}

File: CSharp/ChatApp/AbstractClient.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;
    
namespace ChatApp
{
    public abstract class AbstractClient
    {
        private readonly string SOCKET_ADDRESS = "http://localhost:3001";
        private readonly int TIMEOUT = 10 * 1000 * 60;

        protected HubConnection socket;
        protected ConsoleReader consoleReader = new ConsoleReader();
        protected IdleTimer idleTimer;

        public AbstractClient()
        {
            socket = new HubConnectionBuilder().WithUrl(SOCKET_ADDRESS).Build();
            idleTimer = new IdleTimer(() => Environment.Exit(0), TIMEOUT);
        }

        public abstract Task Start();
    }
}